src/data/finsupp.lean

```diff
@@ -1825,3 +1825,124 @@ lemma finite_lt_nat (n : σ →₀ ℕ) : set.finite {m | m < n} :=
+  intro a, simp only [finsupp.mem_support_iff, ne.def, finset.mem_union, finset.mem_filter],
+  apply and_iff_right_of_imp, contrapose, push_neg,
+  intro h, rw [h.left, h.right], apply inf_idem,
+end
+
+@[simp]
+lemma inf_apply [semilattice_inf β] {a : α} {f g : α →₀ β} : (f ⊓ g) a = f a ⊓ g a := rfl
+
+@[simp]
+lemma nat.support_inf {f g : α →₀ ℕ} : (f ⊓ g).support = f.support ∩ g.support :=
+begin
+  unfold has_inf.inf, dsimp,
+  have h : ∀ m n : ℕ, m ⊓ n = semilattice_inf.inf m n, intros, refl,
+  ext, simp [← h], tauto,
+end
+
+instance has_sup [semilattice_sup β]:
+  has_sup (α →₀ β) :=
+begin
+  refine ⟨_⟩, intros v1 v2,
+  refine ⟨(v1.support ∪ v2.support).filter (λ x, v1 x ⊔ v2 x ≠ 0),
+    (λ (a : α), (v1 a ⊔ v2 a)), _⟩,
+  intro a, simp only [finsupp.mem_support_iff, ne.def, finset.mem_union, finset.mem_filter],
+  apply and_iff_right_of_imp, contrapose, push_neg,
+  intro h, rw [h.left, h.right], apply sup_idem,
+end
+
+@[simp]
+lemma sup_apply [semilattice_sup β] {a : α} {f g : α →₀ β} : (f ⊔ g) a = f a ⊔ g a := rfl
+
+@[simp]
+lemma nat.support_sup {f g : α →₀ ℕ} : (f ⊔ g).support = f.support ∪ g.support :=
+begin
+  unfold has_sup.sup, dsimp, ext,
+  simp only [finsupp.mem_support_iff, ne.def, finset.mem_union, finset.mem_filter],
+  apply and_iff_left_of_imp, repeat {rw ← bot_eq_zero}, contrapose, push_neg,
+  rw ← sup_eq_bot_iff, intro h, rw ← h, refl,
+end
+
+instance lattice : lattice (α →₀ ℕ) :=
+begin
+  refine lattice.mk has_sup.sup has_le.le has_lt.lt _ _ _ _ _ _ _ has_inf.inf _ _ _,
+  exact (finsupp.preorder).le_refl,
+  exact (finsupp.preorder).le_trans,
+  { simp only [auto_param_eq], intros a b, apply lt_iff_le_not_le,
+  },
+  exact (finsupp.partial_order).le_antisymm,
+  intros, rw finsupp.le_iff, intros, simp,
+  intros, rw finsupp.le_iff, intros, simp,
+  { intros, rw finsupp.le_iff at *,
+    intros, rw finsupp.sup_apply, apply lattice.sup_le,
+    { by_cases s ∈ a.support, apply a_1 s h,
+      simp only [finsupp.mem_support_iff, classical.not_not] at h, rw h, simp },
+    { by_cases s ∈ b.support, apply a_2 s h,
+      simp only [finsupp.mem_support_iff, classical.not_not] at h, rw h, simp }
+    },
+  intros, rw finsupp.le_iff, intros, simp,
+  intros, rw finsupp.le_iff, intros, simp,
+  { intros, rw finsupp.le_iff at *, intros,
+    rw finsupp.inf_apply, apply lattice.le_inf, apply a_1 s H, apply a_2 s H }
+end
+
+instance semilattice_inf_bot : semilattice_inf_bot (α →₀ ℕ) :=
+{ bot := 0,
+  bot_le := by { intro a, simp [finsupp.le_iff] },
+..finsupp.lattice}
+
+lemma of_multiset_strict_mono : strict_mono (@finsupp.of_multiset α) :=
+begin
+  unfold strict_mono, intros, rw lt_iff_le_and_ne at *, split,
+  { rw finsupp.le_iff, intros s hs, repeat {rw finsupp.of_multiset_apply},
+    rw multiset.le_iff_count at a_1, apply a_1.left },
+  { have h := a_1.right, contrapose h, simp at *,
+    apply finsupp.equiv_multiset.symm.injective h }
+end
+
+lemma nat.bot_eq_zero : (⊥ : α →₀ ℕ) = 0 := rfl
```

Comment:
This should also move.

Reply:
This one can just be skipped if I provide an instance of canonically_ordered_add_monoid for finsupps to canonically ordered_add_monoids, which I should have done anyway.

Reply:
I'm pretty sure it requires choice though...

Reply:
Mmm, that doesn't quite work, `canonically_ordered_add_monoid`s are slipperier than I thought. I'll remove that part, although maybe there's something about it that could be salvaged in a later PR.

Reply:
I don't think this lemma can actually move, it's not a lemma about `bot` and 0 of type `nat`, but of finsupps. I've added 3 extra versions of it for other similar monoids, although I know that's super awkward.